pyro/infer/rws.py

```diff
@@ -0,0 +1,234 @@
+        as described in [2].
+
+    .. note:: This returns _two_ losses, one each for the model and the guide.
+
+    .. warning:: Mini-batch training is not supported yet.
+
+    :param int num_particles: The number of particles/samples used to form the objective
+        (gradient) estimator. Default is 2.
+    :param insomnia: The scaling between the wake-phi and sleep-phi terms. Default is 1.0 [wake-phi]
+    :param bool model_has_params: Indicate if model has learnable params. Useful in avoiding extra
+        computation when running in pure sleep mode [csis]. Default is True.
+    :param int num_sleep_particles: The number of particles used to form the sleep-phi estimator.
+        Matches `num_particles` by default.
+    :param bool vectorize_particles: Whether the traces should be vectorised
+        across `num_particles`. Default is True.
+    :param int max_plate_nesting: Bound on max number of nested
+        :func:`pyro.plate` contexts. Default is infinity.
+    :param bool strict_enumeration_warning: Whether to warn about possible
+        misuse of enumeration, i.e. that
+        :class:`~pyro.infer.traceenum_elbo.TraceEnum_ELBO` is used iff there
+        are enumerated sample sites.
+
+    References:
+
+    [1] `Reweighted Wake-Sleep`,
+        Jörg Bornschein, Yoshua Bengio
+
+    [2] `Revisiting Reweighted Wake-Sleep for Models with Stochastic Control Flow`,
+        Tuan Anh Le, Adam R. Kosiorek, N. Siddharth, Yee Whye Teh, Frank Wood
+    """
+
+    def __init__(self,
+                 num_particles=2,
+                 insomnia=1.,
+                 model_has_params=True,
+                 num_sleep_particles=None,
+                 vectorize_particles=True,
+                 max_plate_nesting=float('inf'),
+                 max_iarange_nesting=None,  # DEPRECATED
+                 strict_enumeration_warning=True):
+        if max_iarange_nesting is not None:
+            warnings.warn("max_iarange_nesting is deprecated; use max_plate_nesting instead",
+                          DeprecationWarning)
+            max_plate_nesting = max_iarange_nesting
+
+        # force K > 1 otherwise SNIS not possible
+        assert(num_particles > 1), \
+            "Reweighted Wake Sleep needs to be run with more than one particle"
+
+        super(ReweightedWakeSleep, self).__init__(num_particles=num_particles,
+                                                  max_plate_nesting=max_plate_nesting,
+                                                  vectorize_particles=vectorize_particles,
+                                                  strict_enumeration_warning=strict_enumeration_warning)
+        self.insomnia = insomnia
+        self.model_has_params = model_has_params
+        self.num_sleep_particles = num_particles if num_sleep_particles is None else num_sleep_particles
+
+        assert(insomnia >= 0 and insomnia <= 1), \
+            "insomnia should be in [0, 1]"
+
+        print("Running {} {}".format(
```

Comment:
i don't think we want this print statement (we don't do this anywhere else in pyro)

Reply:
done.